features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/forward/ForwardMessagesPresenter.kt

```diff
@@ -0,0 +1,135 @@
+        var results: SearchBarResultState<ImmutableList<RoomSummaryDetails>> by remember { mutableStateOf(SearchBarResultState.NotSearching()) }
+        val forwardingActionState: MutableState<Async<ImmutableList<RoomId>>> = remember { mutableStateOf(Async.Uninitialized) }
+
+        val summaries by client.roomSummaryDataSource.roomSummaries().collectAsState()
+
+        LaunchedEffect(query, summaries) {
+            val filteredSummaries = summaries.filterIsInstance<RoomSummary.Filled>()
+                .map { it.details }
+                .filter { it.name.contains(query, ignoreCase = true) }
+                .toPersistentList()
+            results = if (filteredSummaries.isNotEmpty()) {
+                SearchBarResultState.Results(filteredSummaries)
+            } else {
+                SearchBarResultState.NoResults()
+            }
+        }
+
+        val forwardingSucceeded by remember {
+            derivedStateOf { forwardingActionState.value.dataOrNull() }
+        }
+
+        fun handleEvents(event: ForwardMessagesEvents) {
+            when (event) {
+                is ForwardMessagesEvents.SetSelectedRoom -> {
+                    selectedRooms = persistentListOf(event.room)
+                    // Restore for multi-selection
+//                    val index = selectedRooms.indexOfFirst { it.roomId == event.room.roomId }
+//                    selectedRooms = if (index >= 0) {
+//                        selectedRooms.removeAt(index)
+//                    } else {
+//                        selectedRooms.add(event.room)
+//                    }
+                }
+                ForwardMessagesEvents.RemoveSelectedRoom -> selectedRooms = persistentListOf()
+                is ForwardMessagesEvents.UpdateQuery -> query = event.query
+                ForwardMessagesEvents.ToggleSearchActive -> isSearchActive = !isSearchActive
+                ForwardMessagesEvents.ForwardEvent -> {
+                    isSearchActive = false
+                    val roomIds = selectedRooms.map { it.roomId }.toPersistentList()
+                    matrixCoroutineScope.forwardEvent(eventId, roomIds, forwardingActionState)
+                }
+                ForwardMessagesEvents.ClearError -> forwardingActionState.value = Async.Uninitialized
+            }
+        }
+
+        return ForwardMessagesState(
+            resultState = results,
+            query = query,
+            isSearchActive = isSearchActive,
+            selectedRooms = selectedRooms,
+            isForwarding = forwardingActionState.value.isLoading(),
+            error = (forwardingActionState.value as? Async.Failure)?.error,
+            forwardingSucceeded = forwardingSucceeded,
+            eventSink = { handleEvents(it) }
+        )
+    }
+
+    private fun CoroutineScope.forwardEvent(
+        eventId: EventId,
+        roomIds: ImmutableList<RoomId>,
+        isForwardMessagesState: MutableState<Async<ImmutableList<RoomId>>>,
+    ) = launch {
```

Comment:
Not with the current implementation (if everything goes well, `Result.success(Unit)` is returned, otherwise a `Result.failure(ForwardEventException(failedRoomIds))` is returned instead, but it might not be the best implementation.